File: python-dist/src/qntm/naming.py
"""Local naming store for identities and conversations.

Persists a simple JSON dict at <config_dir>/names.json:
{
    "identities": { "<kid_hex>": "<local_name>", ... },
    "conversations": { "<conv_id_hex>": "<local_name>", ... }
}
"""

import json
import os


class NamingStore:
    """Manage local nicknames for KIDs and conversation IDs."""

    def __init__(self, config_dir: str):
        self._path = os.path.join(config_dir, "names.json")
        self._data = self._load()

    def _load(self) -> dict:
        if os.path.isfile(self._path):
            with open(self._path) as f:
                return json.load(f)
        return {"identities": {}, "conversations": {}}

    def _save(self):
        os.makedirs(os.path.dirname(self._path) or ".", exist_ok=True)
        with open(self._path, "w") as f:
            json.dump(self._data, f, indent=2)
            f.write("\n")

    # --- Identities ---

    def set_identity_name(self, kid_hex: str, name: str):
        """Assign a local name to a KID. Names are unique: if another KID
        already has this name, it loses it."""
        # Remove name from any other KID
        to_remove = [
            k for k, v in self._data["identities"].items()
            if v == name and k != kid_hex
        ]
        for k in to_remove:
            del self._data["identities"][k]
        self._data["identities"][kid_hex] = name
        self._save()

    def resolve_identity_by_name(self, name: str) -> str | None:
        """Look up a KID by its local name. Returns kid_hex or None."""
        for kid, n in self._data["identities"].items():
            if n == name:
                return kid
        return None

    def list_identities(self) -> dict[str, str]:
        """Return {kid_hex: name} for all named identities."""
        return dict(self._data["identities"])

    def remove_identity_name(self, name: str) -> bool:
        """Remove a name by its display name. Returns True if found."""
        for kid, n in list(self._data["identities"].items()):
            if n == name:
                del self._data["identities"][kid]
                self._save()
                return True
        return False

    # --- Conversations ---

    def set_conversation_name(self, conv_id_hex: str, name: str):
        """Assign a local name to a conversation ID. Names are unique."""
        to_remove = [
            k for k, v in self._data["conversations"].items()
            if v == name and k != conv_id_hex
        ]
        for k in to_remove:
            del self._data["conversations"][k]
        self._data["conversations"][conv_id_hex] = name
        self._save()

    def resolve_conversation_by_name(self, name: str) -> str | None:
        """Look up a conversation ID by its local name. Returns hex or None."""
        for cid, n in self._data["conversations"].items():
            if n == name:
                return cid
        return None

    def list_conversations(self) -> dict[str, str]:
        """Return {conv_id_hex: name} for all named conversations."""
        return dict(self._data["conversations"])

    def remove_conversation_name(self, name: str) -> bool:
        """Remove a conversation name. Returns True if found."""
        for cid, n in list(self._data["conversations"].items()):
            if n == name:
                del self._data["conversations"][cid]
                self._save()
                return True
        return False

    # --- Ref resolution helpers ---

    def all_known_ids(self) -> list[str]:
        """Return all known KIDs and conversation IDs."""
        return list(self._data["identities"].keys()) + list(
            self._data["conversations"].keys()
        )

File: attic/go/naming/naming.go
// Package naming implements local address book / nickname system.
// Names are LOCAL ONLY — never transmitted, never shared.
package naming

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sync"
)

// Store is a local name store backed by a JSON file.
type Store struct {
	mu   sync.RWMutex
	path string
	data *storeData
}

type storeData struct {
	// Identity names: hex kid -> local name
	Identities map[string]string `json:"identities"`
	// Conversation names: hex conv_id -> local name
	Conversations map[string]string `json:"conversations"`
}

// NewStore creates or loads a name store at the given path.
func NewStore(dir string) (*Store, error) {
	path := filepath.Join(dir, "names.json")
	s := &Store{
		path: path,
		data: &storeData{
			Identities:    make(map[string]string),
			Conversations: make(map[string]string),
		},
	}
	if data, err := os.ReadFile(path); err == nil {
		if err := json.Unmarshal(data, s.data); err != nil {
			return nil, fmt.Errorf("corrupt names.json: %w", err)
		}
	}
	return s, nil
}

func (s *Store) save() error {
	if err := os.MkdirAll(filepath.Dir(s.path), 0700); err != nil {
		return err
	}
	data, err := json.MarshalIndent(s.data, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(s.path, data, 0600)
}

// SetIdentityName assigns a local name to a kid (hex).
func (s *Store) SetIdentityName(hexKID, name string) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	// Check for name collision
	for k, v := range s.data.Identities {
		if v == name && k != hexKID {
			return fmt.Errorf("name %q already used for %s", name, k)
		}
	}
	s.data.Identities[hexKID] = name
	return s.save()
}

// SetConversationName assigns a local name to a conv_id (hex).
func (s *Store) SetConversationName(hexConvID, name string) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	for k, v := range s.data.Conversations {
		if v == name && k != hexConvID {
			return fmt.Errorf("name %q already used for %s", name, k)
		}
	}
	s.data.Conversations[hexConvID] = name
	return s.save()
}

// RemoveIdentityName removes a name by name string.
func (s *Store) RemoveIdentityName(name string) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	for k, v := range s.data.Identities {
		if v == name {
			delete(s.data.Identities, k)
			return s.save()
		}
	}
	return fmt.Errorf("name %q not found", name)
}

// RemoveConversationName removes a conversation name by name string.
func (s *Store) RemoveConversationName(name string) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	for k, v := range s.data.Conversations {
		if v == name {
			delete(s.data.Conversations, k)
			return s.save()
		}
	}
	return fmt.Errorf("name %q not found", name)
}

// GetIdentityName returns the local name for a kid, or "" if none.
func (s *Store) GetIdentityName(hexKID string) string {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.data.Identities[hexKID]
}

// GetConversationName returns the local name for a conv_id, or "" if none.
func (s *Store) GetConversationName(hexConvID string) string {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.data.Conversations[hexConvID]
}

// ListIdentities returns all identity name mappings.
func (s *Store) ListIdentities() map[string]string {
	s.mu.RLock()
	defer s.mu.RUnlock()
	result := make(map[string]string, len(s.data.Identities))
	for k, v := range s.data.Identities {
		result[k] = v
	}
	return result
}

// ListConversations returns all conversation name mappings.
func (s *Store) ListConversations() map[string]string {
	s.mu.RLock()
	defer s.mu.RUnlock()
	result := make(map[string]string, len(s.data.Conversations))
	for k, v := range s.data.Conversations {
		result[k] = v
	}
	return result
}

// ResolveIdentityByName finds a kid hex by its local name.
func (s *Store) ResolveIdentityByName(name string) (string, bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	for k, v := range s.data.Identities {
		if v == name {
			return k, true
		}
	}
	return "", false
}

// ResolveConversationByName finds a conv_id hex by its local name.
func (s *Store) ResolveConversationByName(name string) (string, bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	for k, v := range s.data.Conversations {
		if v == name {
			return k, true
		}
	}
	return "", false
}
